CDAP-5129 Add entity-id format description to CLI commands

awholegunch opened the pull request:
Fix for https://issues.cask.co/browse/CDAP-5129

Running as a Quick Build: http://builds.cask.co/browse/CDAP-DQB249-5 (this will fail, as develop was broken at the time this branch was created.)

Page of interest: CLI Documentation
http://builds.cask.co/artifact/CDAP-DQB249/shared/build-5/Docs-HTML/4.1.0-SNAPSHOT/en/reference-manual/cli-api.html#available-commands

I have done a pass over all of the documentation strings, and tried to make them as consistent as I could:
- Add quotes to format expression correctly.
- Added single quotes to improve formatting and reading of commands.
- Removed trailing periods on single-sentence descriptions.
- Fixed an error in the CLI rst builder that did not convert an expression enclosed with single-quotes if it began immediately after an opening parentheses.

The "rule" is that anything enclosed in single-quotes in a help description will be converted to a literal in the HTML version of the documentation, so that '\<artifact-id>' will become `'<artifact-id>' `.

Commit:
Add entity-id format description to the CLI commands.

Commit:
Correct the string declaration.

Commit:
Add quotes to format expression correctly.
Added single quotes to improve formatting and reading of commands.
Removed trailing periods on single-sentence descriptions.
Fixed an error in the CLI rst builder that did not convert an expression
enclosed with single-quotes if it began immediately after an opening parentheses.

Commit:
Remove a few more periods.

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/ArgumentName.java:
```diff
@@ -109,14 +109,14 @@
   /**
    * Metadata
    */
+  ENTITY("entity-id"),
```
``entity-id`` is not a metadata-only entity. It is used in other places too, like security

awholegunch replied:
Fixed.

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/metadata/AddMetadataPropertiesCommand.java:
```diff
@@ -34,6 +34,10 @@
   private final MetadataClient client;
 
+  public static final String ENTITY_DESCRIPTION_STRING = String.format("'<%s>' is of the form " +
+    "'<entity-type>:<namespace-id>.<entity-name>', where '<entity-type>' is one of " +
```
we should describe ``<entity-name>`` too, because it can be a composite thing right - e.g. it can be ``<app-name>.<program-type>.<program-name>``

awholegunch replied:
Now we have a problem. We were going to change all `<*-id>` parameters to `<*-name>`...but that would create a problem here. We would end up with `<entity-name>` twice:
```
Gets the metadata of an entity. '<entity-name>' is of the form '<entity-type>:<namespace-id>.<entity-name>', where '<entity-type>' is one of 'artifact', 'app', 'dataset', 'program', 'stream', or 'view'.
```
So, I'm thinking that what we should do is only change from `-id` to `-name` for simple types (non-composite).
The composite types remain as `-id`.

And in this case, we have an `<entity-id>` and a `<program-id>`, where `<program-id>` is composed of
`<app-name>.<program-type>.<program-name>`.

So this code would become:
```
  public static final String ENTITY_DESCRIPTION_STRING = String.format("'<%s>' is of the form " +
    "'<entity-type>:<namespace-id>.<program-id>', where '<entity-type>' is one of " +
    "'artifact', 'app', 'dataset', 'program', 'stream', or 'view', and where '<program-id>' is composed of " +
    "'<app-name>.<program-type>.<program-name>', where '<program-type>' is one of "+
    "flow, mapreduce, service, spark, worker, or workflow." , ENTITY);
```
though is that correct? Should we say `<program-id>` _can_ be either `<program-name>` or composed ... ?

bdmogal replied:
But its not just `<program-id>`. I think the best way to handle this is to list out all the options.
Something like: 
```
public static final String ENTITY_DESCRIPTION_STRING = String.format("'<%s>' is of the form '<entity-type>:<entity-id>', where '<entity-type>' is one of " +
    "'artifact', 'app', 'dataset', 'program', 'stream', or 'view', and where '<entity-id>' uniquely identifies a dataset. For datasets and streams, the <entity-id> is composed of namespace and the entity name, such as <namespace-name>.<dataset-name> or <namespace-name>.<stream-name>. Artifacts and apps have an additional 'version' in the <entity-id>, such as <namespace-name>.<artifact-name>.<artifact-version> and <namespace-name>.<app-name>.<app-version>. Stream Views contain the stream that they were created from - <namespace-name>.<stream-name>.<view-name>, while programs contain the application name and the program type - <namespace-name>.<app-name>.<program-type>.<program-name>, where '<program-type>' is one of flow, mapreduce, service, spark, worker, or workflow.", ENTITY);
```

It may be long, but I think its the only way to describe it without any confusion. Perhaps we could use some line break characters (``\n``) to make this easier on the user.

Ultimately though, we shouldn't be using this internal, encoded string. In future, this should change to 

``add metadata tags <comma-separated-tags> to view <view-name> of stream <stream-name>``; OR

``add metadata tags <comma-separated-tags> to <program-type> <program-name> of application <app-name> [version <app-version>]``.

That's more work, but I think its the only way to make this more readable, with autocomplete. Clearly though, its for future.

awholegunch replied:
What about this (as your example had `entity-id` twice, once as the resolved `ENTITY` and once elsewhere): use `<entity>` instead, and something such as (with some line breaks added here for readability, though they would not appear in the final:
```
public static final String ENTITY_DESCRIPTION_STRING = String.format("

'<entity>' is of the form '<entity-type>:<entity-id>', where '<entity-type>' is one of
'artifact', 'app', 'program', 'dataset', 'stream', or 'view'.

For artifacts and apps, '<entity-id>' is composed of the namespace, entity name, and
version, such as '<namespace-name>.<artifact-name>.<artifact-version>' or
'<namespace-name>.<app-name>.<app-version>'.

For programs, '<entity-id>' includes the application name and the program type:
'<namespace-name>.<app-name>.<program-type>.<program-name>'.

For datasets and streams, '<entity-id>' is the namespace and entity names, such as
'<namespace-name>.<dataset-name>' or '<namespace-name>.<stream-name>'.

For (stream) views, '<entity-id>' includes the stream that they were created from:
'<namespace-name>.<stream-name>.<view-name>'.

'<program-type>' is one of flow, mapreduce, service, spark, worker, or workflow.");
```

bdmogal replied:
sounds good, other than - the line about program type should be right after programs.

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/metadata/AddMetadataPropertiesCommand.java:
```diff
@@ -34,6 +34,10 @@
 
   private final MetadataClient client;
 
+  public static final String ENTITY_DESCRIPTION_STRING = String.format("'<%s>' is of the form " +
```
Doesn't seem like this should be defined here. Its used here as well as some security/authorization commands right? Perhaps define it in ``AbstractCommand`` or similar?

awholegunch replied:
Fixed. Moved to `co.cask.cdap.cli.ArgumentName`.

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/metadata/SearchMetadataCommand.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2016 Cask Data, Inc.
+ * Copyright © 2016-2017 Cask Data, Inc.
```
doesn't seem like this should be updated?

awholegunch replied:
Fixed.

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/command/CreateNamespaceCommand.java:
```diff
@@ -77,6 +77,6 @@ public String getPattern() {
   @Override
   public String getDescription() {
-    return String.format("Creates a %s in CDAP.", ElementType.NAMESPACE.getName());
+    return String.format("Creates a %s in CDAP", ElementType.NAMESPACE.getName());
```
should update copyright?

awholegunch replied:
Fixed.

Commit:
Update copyright strings and move string.

bdmogal commented:
@awholegunch we should add this to the security commands as well.

Commit:
Changed "entity-id" to just "entity". Added usage of the description
string in the security section. Changed order and text for entity
description string block.

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/ArgumentName.java:
```diff
@@ -109,7 +109,7 @@
 
   INSTANCE_URI("cdap-instance-uri"),
   VERIFY_SSL_CERT("verify-ssl-cert"),
-  ENTITY("entity-id"),
```
was this used in a CLI command? If so, this probably breaks compatibility?

awholegunch replied:
I think that it is a placeholder for a value, rather than a value or parameter itself. I will test the change.

awholegunch replied:
I have built and tested this, and it works fine:
```
> get metadata-properties stream:default.john
+==================================================================================================+
| key           | value                                                                            |
+==================================================================================================+
| schema        | {"type":"record","name":"stringBody","fields":[{"name":"body","type":"string"}]} |
| entity-name   | john                                                                             |
| creation-time | 1485994323550                                                                    |
| description   | Meep's friend.                                                                   |
| ttl           | 9223372036854775807                                                              |
+==================================================================================================+
```

bdmogal commented on cdap-cli/src/main/java/co/cask/cdap/cli/ArgumentName.java:
```diff
@@ -140,7 +140,16 @@ public String toString() {
-  public static final String ENTITY_DESCRIPTION_STRING = String.format("'<%s>' is of the form " +
-    "'<entity-type>:<namespace-id>.<entity-name>', where '<entity-type>' is one of " +
-    "'artifact', 'app', 'dataset', 'program', 'stream', or 'view'.", ENTITY);
+  public static final String ENTITY_DESCRIPTION_STRING = String.format("'<%s>' 
```
should we use some ``\n``s to make this easier on the eye?

Commit:
Fix formatted string.

Commit:
attempt to add line breaks.

awholegunch commented:
Closing to move to release/4.1
